fs: fcb: add option to disable CRC for fcb entries

eivindj-nordic opened the pull request:
Add option to disable CRC for fcb entries. This improves the write throughput significantly at the cost of not detecting corrupted data in flash. This is beneficial for aplications that needs the extra write throughput, where error detection is done elsewhere.

Fixes #53707

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>

eivindj-nordic committed:
fs: fcb: add option to disable CRC for fcb entries

Add option to disable CRC for fcb entries. This improves the write
throughput significantly at the cost of not detecting corrupted data
in flash. This is beneficial for aplications that needs the extra
write throughput, where error detection is done elsewhere.

Allow the FCB entries in flash to have a valid CRC when CRC is
disabled in the FCB. This allows existing solutions to disable
CRC checking, while keeping the CRC areas intact. Note that this
is a one-way option.

Fixes #53707

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>

de-nordic commented on subsys/fs/fcb/Kconfig:
```diff
@@ -13,3 +13,13 @@ config FCB
+
+if FCB
+
+config FCB_CRC_DISABLED_ALLOW
```
`FCB_ALLOW_DISABLED_CRC`

de-nordic commented on tests/subsys/fs/fcb/src/main.c:
```diff
@@ -13,7 +13,9 @@
 
-struct fcb test_fcb;
+struct fcb test_fcb = {0};
+struct fcb test_fcb_crc_disabled = { .f_crc_disable = 1 };
```
provide flag definition in fcb.h and macro to set the flag. Usually we do not use the bitfields but directly address bits in flag variable.

de-nordic reviewed (changes requested):
Some comments to address.
Add Release notes update with information of adding the feature and Kconfig option.
Copyright dates update in touched files.

Laczen commented:
@eivindj-nordic, IMO this needs some work. It would be better to provide a config option that enables the crc check (set default to yes, CONFIG_FCB_ENDMARKER_CRC8). And it would be good that the solution would also work for existing fcb areas on devices:
When writing either write the CRC8 or FIXED_END_MARKER depending on the config,
When reading accept CRC8 or CRC8/FIXED_END_MARKER depending on the config

lemrey commented:
I agree that the option should enable the CRC, and be on by default.

eivindj-nordic commented:
Thank you for your feedback! 
> It would be better to provide a config option that enables the crc check (set default to yes, CONFIG_FCB_ENDMARKER_CRC8)

The reason for doing it like this is to make it possible to have FCBs with CRC check enabled and disabled in the same application. Having a Kconfig that enable/disable CRC will apply for all FCBs in the application, but maybe that is ok? I agree that the Kconfig option, as it is per now, can be a bit confusing. Maybe it is ok to drop the Kconfig option and only rely on the flag in `struct fcb`, as not setting the flag would give the original behavior?

> And it would be good that the solution would also work for existing fcb areas on devices: 
When writing either write the CRC8 or FIXED_END_MARKER depending on the config,
When reading accept CRC8 or CRC8/FIXED_END_MARKER depending on the config

Good point! Existing solutions will still work with CRC enabled, but I can rewrite this so that existing solutions can disable CRC if they find it beneficial.

de-nordic commented:
> The reason for doing it like this is to make it possible to have FCBs with CRC check enabled and disabled in the same application. Having a Kconfig that enable/disable CRC will apply for all FCBs in the application

What I think should happen is that when nobody is interested in disabling CRC then it should be possible to not add the feature.
If somebody is interested, then it should be available, and CRC should be Y be default, but you should be able to disable it for a specific  FCB instance, is you use more than one.

eivindj-nordic commented:
> > The reason for doing it like this is to make it possible to have FCBs with CRC check enabled and disabled in the same application. Having a Kconfig that enable/disable CRC will apply for all FCBs in the application
> 
> What I think should happen is that when nobody is interested in disabling CRC then it should be possible to not add the feature. If somebody is interested, then it should be available, and CRC should be Y be default, but you should be able to disable it for a specific FCB instance, is you use more than one.

Ok, I added a Kconfig option, `CONFIG_FCB_FORCE_ENDMARKER_CRC8` that is enabled by default. Name is up for discussion, but I think that should make it clear enough that disabling it still give the possibility to enable CRC if desired.

Laczen commented:
> > The reason for doing it like this is to make it possible to have FCBs with CRC check enabled and disabled in the same application. Having a Kconfig that enable/disable CRC will apply for all FCBs in the application
> 
> What I think should happen is that when nobody is interested in disabling CRC then it should be possible to not add the feature. If somebody is interested, then it should be available, and CRC should be Y be default, but you should be able to disable it for a specific FCB instance, is you use more than one.

I guess this rules out a CONFIG option.

Laczen commented:
> > I guess this rules out a CONFIG option.
> 
> Wait, I am little lost. We are going to have CONFIG, the Kconfig?

What I meant is having a Kconfig option (a `CONFIG_...`) does not play well with a per instance option.

eivindj-nordic commented:
Decided to keep the Kconfig as `CONFIG_FCB_ALLOW_TO_DISABLE_ENDMARKER_CRC` after discussion with @de-nordic.

Laczen commented:
@eivindj-nordic, things that I don't like and need fixing:

a. CONFIG option name, this should be something like `CONFIG_FCB_ALLOW_FIXED_ENDMARKER`,
b. The endmarker should not be `CRC8_CCITT_INITIAL_VALUE`,
c. The swap from `fcb_elem_get_info`: checking a crc to `fcb_elem_get_info` generating a end marker, this is confusing for users that have studied the code in the past,

The solution should be a lot simpler:
`fcb_elem_crc8` is split up in `fcb_elem_crc8()` (does only the calculation) and `fcb_elem_endmarker()` (does checking and sets fixed marker or calls `fcb_elem_crc8()` to set the marker.
`fcb_elem_info` uses `fcb_elem_endmarker()` instead of `fcb_elem_crc8` and compares, if it fails (and the fixed marker is enabled) it calls `fcb_elem_crc8()` and recompares

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -63,14 +65,50 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
 
+int
+fcb_elem_endmarker(struct fcb *_fcb, struct fcb_entry *loc, uint8_t *c8p)
+{
```
start with the check:
```
        if (loc->fe_elem_off + 2 > loc->fe_sector->fs_size) {
		return -ENOTSUP;
	}
```

Laczen replied:
In `fcb_elem_crc8()` this is then no longer required.

eivindj-nordic replied:
true, I kept it as `fcb_elem_crc8` is called other places as well (`fcb_elem_info`), but here it is guarded by calling `fcb_elem_endmarker` first anyway. Will update :)

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -63,14 +65,50 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
+int
+fcb_elem_endmarker(struct fcb *_fcb, struct fcb_entry *loc, uint8_t *c8p)
+{
+#if !IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER)
```
reverse the logic:
```
        if (IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER) && (_fcb->f_flags & FCB_FLAGS_CRC_DISABLED)) {
                *c8p = FCB_FIXED_ENDMARKER;
                return 0;
        }

        return fcb_elem_crc8(_fcb, loc, c8p);
}
```

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -81,6 +119,17 @@ int fcb_elem_info(struct fcb *fcb, struct fcb_entry *loc)
 		return -EIO;
 	}
 
+#if IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER)
```
change this:
```
        if (IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER) && (fl_crc8 != crc8)) {
                rc = fcb_elem_crc8(fcb, loc, &crc8);
                if (rc) {
                        return rc;
                }

        }
        
        if (fl_crc8 != crc8) {
                return -EBADMSG;
        }
```

eivindj-nordic commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -41,6 +41,13 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
 	loc->fe_data_len = len;
 
+#if defined(CONFIG_FCB_CRC_DISABLED_ALLOW)
+	if (fcb->f_crc_disable) {
```
Yes, as we want to allow having both FCBs with CRC enabled and disabled at the same time, so we can't solely rely on the Kconfig.

Laczen replied:
This was some old comment that has been repeated (you can ignore it).

eivindj-nordic commented on subsys/fs/fcb/Kconfig:
```diff
@@ -13,3 +13,13 @@ config FCB
+if FCB
+
+config FCB_CRC_DISABLED_ALLOW
+	bool "Allow disabling Flash Circular Buffer CRC"
```
Should be updated?

Laczen replied:
This was some old comment, you can ignore it. But in general the option should be "positive": `ALLOW_FIXED_ENDMARKER` instead of `ALLOW_DISABLE_CRC`.

eivindj-nordic commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -41,6 +41,13 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
+#if defined(CONFIG_FCB_CRC_DISABLED_ALLOW)
+	if (fcb->f_crc_disable) {
+		*c8p = CRC8_CCITT_INITIAL_VALUE;
+		return 0;
```
This looks outdated? I've added the `#define FCB_FIXED_ENDMARKER 0xab` for this. `~erased_value` might also work.

Laczen commented:
@eivindj-nordic, if I made no mistake the code size with the option disabled (standard) should result in exactly the same code as before inserting the option.

eivindj-nordic commented:
> @eivindj-nordic, if I made no mistake the code size with the option disabled (standard) should result in exactly the same code as before inserting the option. 

@Laczen After addressing your comments, I observe an increase of 4 bytes of flash for my test-application. The FCB flags has to be added to the FCB struct always for 
` if (IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER) && (_fcb->f_flags & FCB_FLAGS_CRC_DISABLED)) {`.
 I think that is ok.  

> And the crc calculation will only be repeated twice in the case of a bad crc and the option fixed marker option disabled for a fcb instance while it is globally enabled.

Yes, there will be some additional processing when CONFIG_FCB_ALLOW_FIXED_ENDMARKER is enabled, and CRC is enabled for the specific FCB. However, it only has to read a couple of bytes extra in that case and not the entire entry content, as is done in` fcb_elem_crc8`. I think that is acceptable when `CONFIG_FCB_ALLOW_FIXED_ENDMARKER` is enabled.

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -63,14 +65,49 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
 {
 	int rc;
-	uint8_t crc8;
+	uint8_t crc8 = FCB_FIXED_ENDMARKER;
```
This is not needed, it will be set by the call to `fcb_elem_endmarker()`

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -63,14 +62,49 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
+	if (IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER) &&
+	    (_fcb->f_flags & FCB_FLAGS_CRC_DISABLED)) {
+
+		uint8_t tmp_str[FCB_TMP_BUF_SZ];
```
It should be possible to move lines 75 - 90 to the start of the routine and not to duplicate this in fcb_elem_crc8 (the tmp_str to calculate the len is only a 2 byte string).

eivindj-nordic replied:
Ok, I've restructured the `fcb_elem_endmarker` and  `fcb_elem_crc8` to remove the `tmp_str` from `fcb_elem_crc8` and to remove the flag in` fcb` struct when `CONFIG_FCB_ALLOW_FIXED_ENDMARKER` is not enabled. Application size with `CONFIG_FCB_ALLOW_FIXED_ENDMARKER` disabled is now the same as without this PR.

Laczen commented:
> > @eivindj-nordic, if I made no mistake the code size with the option disabled (standard) should result in exactly the same code as before inserting the option.
> 
> @Laczen After addressing your comments, I observe an increase of 4 bytes of flash for my test-application. The FCB flags has to be added to the FCB struct always for ` if (IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER) && (_fcb->f_flags & FCB_FLAGS_CRC_DISABLED)) {`. I think that is ok.
> 
It should be possible not to have this (maybe by splitting this into two if's). I thought the compiler was smarter 😀

Laczen commented on include/zephyr/fs/fcb.h:
```diff
@@ -129,6 +135,11 @@ struct fcb {
 	 * flash parameters and initialized upon call to fcb_init.
 	 */
+
+#if IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER)
```
In include files you can use `#ifdef`.
In c files where it is used as part of a routine it is better to use `if IS_ENABLED()` to improve readability of your code

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -1,58 +1,43 @@
-int
-fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
+static int
+fcb_elem_crc8(struct fcb *_fcb, struct fcb_entry *loc, uint8_t *tmp_str, size_t tmp_str_len,
```
IMO passing the `tmp_str` is a little too much trying to get the code size equal. The function should be calculate the crc8 given a "correct" `fcb_entry`. It is cleaner to define a `len_str` when checking the `fcb_entry` and then a `buf[FCB_TMP_BUF_SZ]` when calculating the crc8.

Laczen commented on subsys/fs/fcb/fcb_elem_info.c:
```diff
@@ -63,25 +48,78 @@ fcb_elem_crc8(struct fcb *fcb, struct fcb_entry *loc, uint8_t *c8p)
-	uint8_t crc8;
-	uint8_t fl_crc8;
+
+	(void)fixed_endmarker_allow;
```
What does this do ?

eivindj-nordic replied:
It is removing a unused variable compilation warning if `CONFIG_FCB_ALLOW_FIXED_ENDMARKER` is not enabled

Laczen reviewed (dismissed):
A small nitpick, with seperate crc8/fixed em calculation the if from line 29 (original fcb_elem_info.c) would better be repeated in each separate calculation.

eivindj-nordic commented:
> fcb_elem_info.c

Updated. Thank you for your patience :)
